torch/distributed/algorithms/model_averaging/utils.py

```diff
@@ -19,18 +19,45 @@ def average_parameters(
     if dist._rank_not_in_group(group_to_use):
         return
 
+    if isinstance(params, types.GeneratorType):
+        # compatible with model.parameters() input
+        flat_params_all_reduce(params, group_to_use)
+    elif isinstance(params, list):
```

Comment:
This is essentially parameter groups stored in an optimizer. I think it's better to hide this complexity in `PostLocalSGDOptimizer` rather than this lower level.

Reply:
agree.
i will fix later

Reply:
Please see my new comment here: https://github.com/pytorch/pytorch/pull/74894#discussion_r841193929

I think it makes sense to handle parameter groups here. Otherwise as you pointed out, it would not be efficient during the warmup stage.

Reply:
see new comment. https://github.com/pytorch/pytorch/pull/74894#discussion_r842465678